File: 53.Bellman–FordAlgo.java
import java.io.*;
import java.util.*;

/*
 * Bellman-Ford Algorithm in Java
 * Finds shortest paths from a single source, detects negative-weight cycles
 */

public class BellmanFordAlgo {

    // Class to represent an Edge (u -> v with weight w)
    static class Edge {
        int u, v, w;
        Edge(int u, int v, int w) {
            this.u = u;
            this.v = v;
            this.w = w;
        }
    }

    // Bellman-Ford implementation
    static void bellmanFord(List<Edge> edges, int V, int source) {
        // Step 1: Initialize distances with "infinity"
        int[] dist = new int[V];
        int[] parent = new int[V]; // to reconstruct path
        Arrays.fill(dist, Integer.MAX_VALUE);
        Arrays.fill(parent, -1);

        // Distance to source is zero
        dist[source] = 0;

        // Step 2: Relax all edges (V-1) times
        for (int i = 1; i < V; i++) {
            boolean updated = false; // flag to detect early termination
            for (Edge e : edges) {
                if (dist[e.u] != Integer.MAX_VALUE && dist[e.u] + e.w < dist[e.v]) {
                    dist[e.v] = dist[e.u] + e.w;
                    parent[e.v] = e.u;
                    updated = true;
                }
            }
            if (!updated) break; // No updates → already optimal
        }

        // Step 3: Check for negative-weight cycles
        boolean hasNegCycle = false;
        for (Edge e : edges) {
            if (dist[e.u] != Integer.MAX_VALUE && dist[e.u] + e.w < dist[e.v]) {
                hasNegCycle = true;
                break;
            }
        }

        // Step 4: Output results
        if (hasNegCycle) {
            System.out.println("Negative cycle detected! Shortest paths are undefined.");
            return;
        }

        System.out.println("Vertex | Distance from Source | Path");
        for (int v = 0; v < V; v++) {
            System.out.printf("%6d | %20s | ", v,
                (dist[v] == Integer.MAX_VALUE ? "∞" : dist[v]));
            printPath(parent, v);
            System.out.println();
        }
    }

    // Helper function to print reconstructed path
    static void printPath(int[] parent, int v) {
        if (v == -1) return;
        List<Integer> path = new ArrayList<>();
        int cur = v;
        Set<Integer> seen = new HashSet<>(); // avoid infinite loop
        while (cur != -1) {
            if (seen.contains(cur)) { // detects cycle in parent chain
                System.out.print("(cycle)");
                return;
            }
            seen.add(cur);
            path.add(cur);
            cur = parent[cur];
        }
        Collections.reverse(path);
        for (int i = 0; i < path.size(); i++) {
            System.out.print(path.get(i));
            if (i < path.size() - 1) System.out.print("->");
        }
    }

    public static void main(String[] args) {
        // Example 1: Graph without a negative cycle
        int V = 5;
        List<Edge> edges = new ArrayList<>();

        // Define edges (u, v, w)
        edges.add(new Edge(0, 1, 6));
        edges.add(new Edge(0, 2, 7));
        edges.add(new Edge(1, 2, 5));
        edges.add(new Edge(1, 3, -4));
        edges.add(new Edge(2, 4, -3));
        edges.add(new Edge(4, 3, 7));

        System.out.println("---- Run 1: No Negative Cycle ----");
        bellmanFord(edges, V, 0);

        // Example 2: Add an edge that introduces a negative cycle
        edges.add(new Edge(3, 1, -10)); // creates cycle 1->3->1 with -14 total

        System.out.println("\n---- Run 2: With Negative Cycle ----");
        bellmanFord(edges, V, 0);
    }
}

File: 53.Bellman–FordAlgo.py
# Bellman–Ford is a single-source shortest-path algorithm that handles graphs with 
# negative edge weights and detects negative-weight cycles reachable from the source.

# It runs in O(V + E) 

# O(V⋅E) time (V = vertices, E = edges) and returns shortest distances or
# reports that shortest paths are undefined when a negative cycle is reachable.

# Explanation

# Bellman–Ford repeatedly relaxes every edge up to V - 1 times 
# (each pass can propagate shortest-path information across one more edge).

# After V - 1 relaxations, if any edge can still be 
# relaxed, then there is a negative-weight cycle reachable from the source 
# (because distances can be driven down indefinitely).

# Use Bellman–Ford when your graph may contain negative-weight edges 
# (but not for dense graphs where faster algorithms for non-negative 
#  weights such as Dijkstra would be preferable).

# Time complexity: O(V . E) Space: O(V) for distances and parent pointers.

#  Finds shortest paths from a single source and detects negative-weight cycles.

#        Graph Explanation:

#        0 -> 1 (weight 6)
#        0 -> 2 (weight 7)
#        1 -> 2 (weight 5)
#        1 -> 3 (weight -4)
#        2 -> 4 (weight -3)
#        4 -> 3 (weight 7)

#    Graph Representation:

#          (6)
#      (0) -----> (1)
#       | \        |  \
#      (7) \       |   \ (-4)
#           \      |    \
#            > (2) |     > (3)
#             \(-3)| 
#              \   |
#               > (4)
#                (7)

def bellman_ford(edges, V, source):
    dist = [float('inf')] * V
    dist[source] = 0
    parent = [None] * V

    for i in range(V - 1):
        updated = False
        for(u, v, w) in edges:
            if dist[u] != float('inf') and dist[u] + w < dist[v]:
                dist[v] = dist[u] + w
                parent[v] = u
                updated = True

        if not updated:
            break

    negative_cycle = False
    for (u, v, w) in edges:
        if dist[u] != float('inf') and dist[u] + w < dist[v]:
            negative_cycle = True
            break

    return dist, parent, negative_cycle

def reconstruct_path(parent, dest, V):
    path = []
    curr = dest
    visited = set()
    while curr is not None:
        if curr in visited:
            return None
        visited.add(curr)
        path.append(curr)
        curr = parent[curr]
    path.reverse()
    return path

# -------------------------
# Example 1: No negative cycle
# -------------------------
V = 5  # nodes 0..4
edges = [
    (0, 1, 6),   # 0 -> 1 (6)
    (0, 2, 7),   # 0 -> 2 (7)
    (1, 2, 5),   # 1 -> 2 (5)
    (1, 3, -4),  # 1 -> 3 (-4) negative edge but no 3->1 yet (no cycle)
    (2, 4, -3),  # 2 -> 4 (-3)
    (4, 3, 7)    # 4 -> 3 (7)
]

dist, parent, neg = bellman_ford(edges, V, 0)

print("---- Run 1: No negative cycle ----")
print(f"Negative cycle detected? {neg}")
if not neg:
    print("Vertex | Distance from 0 | Shortest path from 0")
    for v in range(V):
        if dist[v] == float('inf'):
            dstr = "∞"
            path_str = "unreachable"
        else:
            dstr = str(dist[v])
            p = reconstruct_path(parent, v, V)
            path_str = "(path undefined)" if p is None else "->".join(map(str, p))
        print(f"{v:>6} | {dstr:>15} | {path_str}")
else:
    print("Cannot display paths because a negative-weight cycle is present.")

# -------------------------
# Example 2: Introduce a negative cycle
# -------------------------
# Add an edge that creates a negative cycle: 1 -> 3 (-4) and 3 -> 1 (-10) -> cycle weight -14
edges_with_cycle = edges + [(3, 1, -10)]

dist2, parent2, neg2 = bellman_ford(edges_with_cycle, V, 0)

print("\n---- Run 2: With a negative cycle added (3 -> 1 : -10) ----")
print(f"Negative cycle detected? {neg2}")
if not neg2:
    print("Vertex | Distance from 0 | Shortest path from 0")
    for v in range(V):
        if dist2[v] == float('inf'):
            dstr = "∞"
            path_str = "unreachable"
        else:
            dstr = str(dist2[v])
            p = reconstruct_path(parent2, v, V)
            path_str = "(path undefined)" if p is None else "->".join(map(str, p))
        print(f"{v:>6} | {dstr:>15} | {path_str}")
else:
    print("Since a negative-weight cycle is reachable from the source, shortest paths are undefined (can decrease indefinitely).")